# Running Kallisto long read version

Please make sure that you already installed neccassary packages.

## Making indices

```bash
#!/bin/sh
#SBATCH -A model-ad_lab
#SBATCH --cpus-per-task 10
#SBATCH --output=kallisto_index_LR.out
#SBATCH --error=kallisto_index_LR.err
#SBATCH --time=1:00:00
#SBATCH -J kallisto_index_LR
#SBATCH --mail-type=START,END
#SBATCH --partition=standard

ref='name_refrence_genome #mm39
ref_genome='path_to_refrence_genome' #mm39.fa.gz
ref_annot='path_to_gencode_gtf' #gencode.vM32.chr_patch_hapl_scaff.annotation.gtf.gz

path_to_kallisto='kallisto/build/src/kallisto'
path_to_bustools='bustools/build/src/bustools'

#build index
kb ref --kallisto ${path_to_kallisto} \
-i ${ref}_k-63.idx \
-k 63 \
-f1 ${ref}.cdna.fa \
-g ${ref}.t2g.txt \
${ref_genome} \
${ref_annot} \
--overwrite

```

**Note**: Indices generated with kallisto are not compatible with indices generated by lr-kallisto, due to the way that longer k-mers (>31) must be stored in 16 bytes of memory vs the 8 bytes used for kallisto.

**More information**:
- [kallisto overview](https://pachterlab.github.io/kallisto/starting.html)
- [kallisto transcriptome indices](https://github.com/pachterlab/kallisto-transcriptome-indices)


## Quantification

### Run Kallisto on bulk long-read data

This script will generate and submit separete jobs for each sample to run lr-Kallisto on bulk long read data to generate transcript-level count and tpm matrices: 

```bash
#!/bin/sh
#SBATCH -A model-ad_lab
#SBATCH --cpus-per-task 1
#SBATCH --output=kallisto.out
#SBATCH --error=kallisto.err
#SBATCH --time=1:00:00
#SBATCH -J kallisto_lr
#SBATCH --mail-type=START,END
#SBATCH --partition=standard

ref='name_refrence_genome #mm39
ref_genome='path_to_refrence_genome' #mm39.fa.gz
ref_annot='path_to_gencode_gtf' #gencode.vM32.chr_patch_hapl_scaff.annotation.gtf.gz
reads='path_to_reads_file' #"/share/crsp/lab/model-ad/nargesr/kallisto_lr/AD003_Sep24_mm39/fastqs"

path_to_kallisto='kallisto/build/src/kallisto'
path_to_bustools='bustools/build/src/bustools'

output='path_to_output' #'/share/crsp/lab/model-ad/nargesr/kallisto_lr/AD003_Sep24_mm39/output'

#pseudoalign reads
COUNTER=1
IFS=$'\n'
for line in $(cat sample_name.txt)
do
    # Use awk to parse the columns from each line
    sample_name=$(echo $line | awk '{print $1}')
    fastqs_num=$(echo $line | awk '{print $2}')
    fastqs=()
    for i in $(seq 1 $fastqs_num);
    do
      	fastq_index=$((i+2))
        fastqs+=(${reads}/$(echo $line | awk -v idx=$fastq_index '{print $idx}'))
    done
    printf "Sample: %s\n" "$sample_name"
    printf "Fastqs: %s\n" "${fastqs[@]}"

    # Combine all fastqs into a single string with spaces
    fastqs_string=$(printf "%s " "${fastqs[@]}")
    
    # Remove trailing space
    fastqs_string=$(echo $fastqs_string | sed 's/ *$//')
    printf "cat $fastqs_string > ${reads}/${sample_name}.fastq.gz\n"
    cat "${fastqs[@]}" > ${reads}/${sample_name}.fastq.gz

    scriptName=${sample_name}
    curr=${sample_name}.sh
    output_sample=${output}_${sample_name}
    fastq_file=${reads}/${sample_name}.fastq.gz
    
    echo '#!/bin/bash' > ${curr}
    echo '#SBATCH -A model-ad_lab' >> ${curr}
    echo '#SBATCH --cpus-per-task 16' >> ${curr}
    echo '#SBATCH --output=kallisto_%J.out' >> ${curr}
    echo '#SBATCH --error=kallisto_%J.err' >> ${curr}
    echo '#SBATCH --time=02:00:00' >> ${curr}
    echo '#SBATCH -J kallisto_%J' >> ${curr}
    echo '#SBATCH --mail-type=START,END' >> ${curr}
    echo '#SBATCH --partition=standard' >> ${curr}

    echo "ref='name_refrence_genome'" >> ${curr}
    echo "reads='path_to_reads_file'" >> ${curr}
    echo "path_to_kallisto='kallisto/build/src/kallisto'" >> ${curr}
    echo "path_to_bustools='bustools/build/src/bustools'" >> ${curr}
    echo "output='${output_sample}'" >> ${curr}
    
    echo "${path_to_kallisto} bus --long --threshold 0.8 -x bulk -i ${ref}_k-63.idx -o ${output_sample} ${fastq_file} ${name} -t 8" >> ${curr}

    echo "${path_to_bustools} sort -t 32 ${output_sample}/output.bus -o ${output_sample}/sorted.bus" >> ${curr} 
    echo "${path_to_bustools} count ${output_sample}/sorted.bus -t ${output_sample}/transcripts.txt  -e ${output_sample}/matrix.ec  -o ${output_sample}/count --cm -m -g ${ref}.t2g.txt" >> ${curr}

    echo "${path_to_kallisto} quant-tcc -t 32 --long -P ONT ${output_sample}/count.mtx -i ${ref}_k-63.idx -f ${output_sample}/flens.txt -e ${output_sample}/count.ec.txt -o ${output_sample}" >> ${curr}
    
    chmod +x ${curr}
    sbatch ${curr}

    COUNTER=`expr $COUNTER + 1`
    
done


```

`sample_name.txt` used in the script contains the name of the sample followed by the number of fastq files we have for the samples and then the corresponding fastq files in each line and it should look like this:

```
ad003_11616_lig-blk   2   ad003_11616_lig-blk_1.fastq.gz   ad003_11616_lig-blk_2.fastq.gz
ad003_11617_lig-blk   2   ad003_11617_lig-blk_1.fastq.gz   ad003_11617_lig-blk_2.fastq.gz
ad003_11625_lig-blk   2   ad003_11625_lig-blk_1.fastq.gz   ad003_11625_lig-blk_2.fastq.gz
ad003_11627_lig-blk   2   ad003_11627_lig-blk_1.fastq.gz   ad003_11627_lig-blk_2.fastq.gz
ad003_11628_lig-blk   2   ad003_11628_lig-blk_1.fastq.gz   ad003_11628_lig-blk_2.fastq.gz
ad003_11629_lig-blk   2   ad003_11629_lig-blk_1.fastq.gz   ad003_11629_lig-blk_2.fastq.gz
ad003_12517_lig-blk   2   ad003_12517_lig-blk_1.fastq.gz   ad003_12517_lig-blk_2.fastq.gz
ad003_12648_lig-blk   2   ad003_12648_lig-blk_1.fastq.gz   ad003_12648_lig-blk_2.fastq.gz
ad003_12649_lig-blk   2   ad003_12649_lig-blk_1.fastq.gz   ad003_12649_lig-blk_2.fastq.gz
ad003_12659_lig-blk   2   ad003_12659_lig-blk_1.fastq.gz   ad003_12659_lig-blk_2.fastq.gz
ad003_12660_lig-blk   2   ad003_12660_lig-blk_1.fastq.gz   ad003_12660_lig-blk_2.fastq.gz
ad003_12670_lig-blk   2   ad003_12670_lig-blk_1.fastq.gz   ad003_12670_lig-blk_2.fastq.gz
```

In the end, the structure of your files would be something like this:

````
├── ad003_11517_lig-blk.sh
├── ad003_11616_lig-blk.sh
├── ad003_11617_lig-blk.sh
├── ad003_11625_lig-blk.sh
├── ad003_11627_lig-blk.sh
├── ad003_11628_lig-blk.sh
├── ad003_11629_lig-blk.sh
├── ad003_11648_lig-blk.sh
├── ad003_11649_lig-blk.sh
├── ad003_11659_lig-blk.sh
├── ad003_11660_lig-blk.sh
├── ad003_11670_lig-blk.sh
├── ad003_12517_lig-blk.sh
├── ad003_12648_lig-blk.sh
├── ad003_12649_lig-blk.sh
├── ad003_12659_lig-blk.sh
├── ad003_12660_lig-blk.sh
├── ad003_12670_lig-blk.sh
├── fastqs
│   ├── ad003_11616_lig-blk_1.fastq.gz
│   ├── ad003_11616_lig-blk_2.fastq.gz
│   ├── ad003_11617_lig-blk_1.fastq.gz
│   ├── ad003_11617_lig-blk_2.fastq.gz
│   ├── ad003_11625_lig-blk_1.fastq.gz
│   ├── ad003_11625_lig-blk_2.fastq.gz
│   ├── ad003_11627_lig-blk_1.fastq.gz
│   ├── ad003_11627_lig-blk_2.fastq.gz
│   ├── ad003_11628_lig-blk_1.fastq.gz
│   ├── ad003_11628_lig-blk_2.fastq.gz
│   ├── ad003_11629_lig-blk_1.fastq.gz
│   ├── ad003_11629_lig-blk_2.fastq.gz
│   ├── ad003_12517_lig-blk_1.fastq.gz
│   ├── ad003_12517_lig-blk_2.fastq.gz
│   ├── ad003_12648_lig-blk_1.fastq.gz
│   ├── ad003_12648_lig-blk_2.fastq.gz
│   ├── ad003_12649_lig-blk_1.fastq.gz
│   ├── ad003_12649_lig-blk_2.fastq.gz
│   ├── ad003_12659_lig-blk_1.fastq.gz
│   ├── ad003_12659_lig-blk_2.fastq.gz
│   ├── ad003_12660_lig-blk_1.fastq.gz
│   ├── ad003_12660_lig-blk_2.fastq.gz
│   ├── ad003_12670_lig-blk_1.fastq.gz
│   └── ad003_12670_lig-blk_2.fastq.gz
├── output_ad003_11616_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11617_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11625_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11627_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11628_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11629_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12517_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12648_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12649_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12659_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12660_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12670_lig-blk
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── run.sh
└── sample_name.txt
````
