# Pipeline for analysing lond read data using long read Kallisto

## setting up and install neccessary packages

### Install LR kallisto

LR-kallisto is located in [this repo](https://github.com/bound-to-love/kallisto.git).

Follow commands for installation

```bash
git clone https://github.com/bound-to-love/kallisto.git
cd kallisto
mkdir build
cd build
cmake ..
make 
```
### Install other neccessary pakcages

1. [bustools](https://github.com/BUStools/bustools?tab=readme-ov-file#installation)
2. [kb-python](https://github.com/pachterlab/kb_python?tab=readme-ov-file#installation)



## Making indices

```bash
#!/bin/sh
#SBATCH -A model-ad_lab
#SBATCH --cpus-per-task 10
#SBATCH --output=kallisto_index_LR.out
#SBATCH --error=kallisto_index_LR.err
#SBATCH --time=1:00:00
#SBATCH -J kallisto_index_LR
#SBATCH --mail-type=START,END
#SBATCH --partition=standard

ref='path_to_refrence_genome'

path_to_lr_kallisto='kallisto/build/src/kallisto'

#build index
kb ref --kallisto ${path_to_lr_kallisto} -i ${ref}_k-63.idx -k 63 -f1 ${ref}.cdna.fa -g ${ref}.t2g.txt ${ref}.genome.fa.gz ${ref}.annotation.gtf.gz --overwrite

```

**Note**: Indices generated with kallisto are not compatible with indices generated by lr-kallisto, due to the way that longer k-mers (>31) must be stored in 16 bytes of memory vs the 8 bytes used for kallisto.

**More information**:
- [kallisto overview](https://pachterlab.github.io/kallisto/starting.html)
- [kallisto transcriptome indices](https://github.com/pachterlab/kallisto-transcriptome-indices)


## Quantification

### Run lr-kallisto on bulk long read data

This script will generate and submitt separete jobs for each sample to run lr-kallisto on bulk long read data to generate transcript-level count and tpm matrices: 

```bash
#!/bin/sh
#SBATCH -A model-ad_lab
#SBATCH --cpus-per-task 1
#SBATCH --output=kallisto.out
#SBATCH --error=kallisto.err
#SBATCH --time=1:00:00
#SBATCH -J kallisto_pseudoalign_run
#SBATCH --mail-type=START,END
#SBATCH --partition=standard

ref='path_to_refrence_genome'
reads='path_to_reads_file'
path_to_lr_kallisto='kallisto/build/src/kallisto'
path_to_bustools='bustools/build/src/bustools'
output='path_to_output'

#pseudoalign reads
COUNTER=1
IFS=$'\n'
for sample in $(cat sample_name.txt)
do
    scriptName=${sample}
    curr=${sample}.sh
    output_sample=${output}_${sample}
    fastq_file=${reads}/${sample}.fastq
    
    echo '#!/bin/bash' > ${curr}
    echo '#SBATCH -A model-ad_lab' >> ${curr}
    echo '#SBATCH --cpus-per-task 16' >> ${curr}
    echo '#SBATCH --output=kallisto_%J.out' >> ${curr}
    echo '#SBATCH --error=kallisto_%J.err' >> ${curr}
    echo '#SBATCH --time=02:00:00' >> ${curr}
    echo '#SBATCH -J kallisto_%J' >> ${curr}
    echo '#SBATCH --mail-type=START,END' >> ${curr}
    echo '#SBATCH --partition=standard' >> ${curr}

    echo "ref='path_to_refrence_genome'" >> ${curr}
    echo "reads='path_to_reads_file'" >> ${curr}
    echo "path_to_lr_kallisto='kallisto/build/src/kallisto'" >> ${curr}
    echo "path_to_bustools='bustools/build/src/bustools'" >> ${curr}
    echo "output='${output_sample}'" >> ${curr}
    
    echo "${path_to_lr_kallisto} bus --long --threshold 0.8 -x bulk -i ${ref}_k-63.idx -o ${output_sample} ${fastq_file} ${name} -t 8" >> ${curr}

    echo "${path_to_bustools} sort -t 32 ${output_sample}/output.bus -o ${output_sample}/sorted.bus" >> ${curr} 
    echo "${path_to_bustools} count ${output_sample}/sorted.bus -t ${output_sample}/transcripts.txt  -e ${output_sample}/matrix.ec  -o ${output_sample}/count --cm -m -g ${ref}.t2g.txt" >> ${curr}

    echo "${path_to_lr_kallisto} quant-tcc -t 32 --long -P ONT ${output_sample}/count.mtx -i ${ref}_k-63.idx -f ${output_sample}/flens.txt -e ${output_sample}/count.ec.txt -o ${output_sample}" >> ${curr}
    
    chmod +x ${curr}
    sbatch ${curr}

    COUNTER=`expr $COUNTER + 1`
    
done
```

`sample_name.txt` used in the script contains name of combined trimmed fastq at sample level and it should be look like this:

```
ad003_11616_lig-blk_t1
ad003_11617_lig-blk_t1
ad003_11625_lig-blk_t1
ad003_11627_lig-blk_t1
ad003_11628_lig-blk_t1
ad003_11629_lig-blk_t1
ad003_12517_lig-blk_t1
ad003_12648_lig-blk_t1
ad003_12649_lig-blk_t1
ad003_12659_lig-blk_t1
ad003_12660_lig-blk_t1
ad003_12670_lig-blk_t1
```

At the end, the structure of your files would be something like this:

````
├── ad003_11616_lig-blk_t1.sh
├── ad003_11617_lig-blk_t1.sh
├── ad003_11625_lig-blk_t1.sh
├── ad003_11627_lig-blk_t1.sh
├── ad003_11628_lig-blk_t1.sh
├── ad003_11629_lig-blk_t1.sh
├── ad003_12517_lig-blk_t1.sh
├── ad003_12648_lig-blk_t1.sh
├── ad003_12649_lig-blk_t1.sh
├── ad003_12659_lig-blk_t1.sh
├── ad003_12660_lig-blk_t1.sh
├── ad003_12670_lig-blk_t1.sh
├── output_ad003_11616_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11617_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11625_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11627_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11628_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_11629_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12517_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12648_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12649_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12659_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12660_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── output_ad003_12670_lig-blk_t1
│   ├── count.barcodes.txt
│   ├── count.ec.txt
│   ├── count.mtx
│   ├── flens.txt
│   ├── index.saved
│   ├── matrix.abundance.mtx
│   ├── matrix.abundance.tpm.mtx
│   ├── matrix.cells
│   ├── matrix.ec
│   ├── matrix.efflens.mtx
│   ├── matrix.fld.tsv
│   ├── matrix.sample.barcodes
│   ├── novel.fastq
│   ├── output.bus
│   ├── run_info.json
│   ├── sorted.bus
│   ├── transcript_lengths.txt
│   └── transcripts.txt
├── run.sh
├── sample_name.txt
├── trimcombinedfastq
│   ├── ad003_11616_lig-blk_t1.fastq
│   ├── ad003_11617_lig-blk_t1.fastq
│   ├── ad003_11625_lig-blk_t1.fastq
│   ├── ad003_11627_lig-blk_t1.fastq
│   ├── ad003_11628_lig-blk_t1.fastq
│   ├── ad003_11629_lig-blk_t1.fastq
│   ├── ad003_12517_lig-blk_t1.fastq
│   ├── ad003_12648_lig-blk_t1.fastq
│   ├── ad003_12649_lig-blk_t1.fastq
│   ├── ad003_12659_lig-blk_t1.fastq
│   ├── ad003_12660_lig-blk_t1.fastq
│   └── ad003_12670_lig-blk_t1.fastq
````

### Concate count and tom matrices

Once you have transcript-level count and tpm matrices for each sample, it's time to concat them. Please look at the (this notebook)[analysis_pipeline.ipynb] for more information.